### ISSUE TITLE
Resnet20: add fake quantization nodes

### ISSUE THREAD

**Issue by mynameischaos** [tensorflow/tensorflow#20141]
System information
------------------------------------------
Have I written custom code (as opposed to using a stock example script provided in TensorFlow):no
OS Platform and Distribution (e.g., Linux Ubuntu 16.04):CentOS Linux release 7.2
TensorFlow installed from (source or binary):Anaconda python 3.6.5(conda install)
TensorFlow version (use command below):1.8.0
Python version: 3.6.5
Bazel version (if compiling from source): 0.7.0
GCC/Compiler version (if compiling from source): 4.8.5
CUDA/cuDNN version:cuda-9.0
GPU model and memory: P40
Phone: N/A

Hi~ 
When I use 'tf.contrib.quantize.experimental_create_training_graph'  it will add fake quantization in the graph.  reference this picture(find positions to insert fake quantization nodes). 
![image](https://user-images.githubusercontent.com/9084403/41646572-c5a48d8a-74a6-11e8-81bf-d1fee67d5c4a.png)

I think it will add a fake quantization node in the bypass, but I can not find it in the tensorboard graph(in the red box, it is the bypass). In other convs, I can find the quantization node. 
![image](https://user-images.githubusercontent.com/9084403/41649732-d6f538d4-74ae-11e8-87b5-66b2b49904a7.png)

My network architecture is ResNet20,  tensorflow version is 1.8.0

pls help me. thank you.

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce

**Comment by suharshs** [comment]
Is this ResnetV2. If that is the case, there are issues with batchnorm that messes up quantization, that we still need to resolve before we are able to quantize this.

**Comment by mynameischaos** [comment]
@suharshs  When will this problem be solved?

**Comment by suharshs** [comment]
This may be a bit involved to resolve, but it is something we plan to look into this upcoming quarter.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @suharshs: It has been 17 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @suharshs: It has been 31 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @suharshs, @raghuraman-k: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by calicratis19** [comment]
I'm getting same type of error for xception model.
` Array Cast, which is an input to the ResizeBilinear operator producing the output array ResizeBilinear, is lacking min/max data, which is necessary for quantization. If accuracy matters, either target a non-quantized output format, or run quantized training with your model from a floating point checkpoint to change the input graph to contain min/max information. If you don't care about accuracy, you can pass --default_ranges_min= and --default_ranges_max= for easy experimentation.
`
The model is downloaded from tensorflow deeplab repo.

The command that generated the error was

`bazel-bin/tensorflow/contrib/lite/toco/toco \
--input_file=frozen_inference_graph.pb \
--input_format=TENSORFLOW_GRAPHDEF \
--output_format=TFLITE \
--output_file=xception.lite \
--inference_type=QUANTIZED_UINT8 \
--input_arrays=ImageTensor \
--output_arrays=SemanticPredictions \
--input_shapes=1,299,299,3`

**Comment by suharshs** [comment]
The issue is that the contrib/quantize rewriter is not very robust to any arbitrary model yet.

It can be complicated and sometimes not possible to fully quantize certain models due to the available fused operations at any given time , if you goal is to just get a smaller and faster model, I recommend trying the --post_training_quantize flag to tflite_convert. With that you keep the inference_type=FLOAT and pass a floating point version of your model (no need to call the contrib/quantize tool). That may provide sufficient speedup for your use case. Check it out here: https://www.tensorflow.org/performance/post_training_quantization

**Comment by tensorflowbutler** [comment]
Nagging Assignees @suharshs, @raghuraman-k: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @suharshs, @raghuraman-k: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @suharshs: It has been 45 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @suharshs: It has been 60 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @suharshs: It has been 75 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by suharshs** [comment]
Based on feedback that the contrib/quantize quantization-aware training tool is a bit brittle and hard to use on some model architectures, we have released a [post-training integer quantization tool](https://medium.com/tensorflow/tensorflow-model-optimization-toolkit-post-training-integer-quantization-b4964a1ea9ba), that requires a small calibration dataset. Please take a look at the [tutorial](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/tutorials/post_training_integer_quant.ipynb) and give it a try, it should work much better! And let us know if you run into any issues.

Closing this issue, since we are rethinking and working on an api to replace contrib/quantize quantization-aware-training (although post-training quantization above should be sufficient for the majority of use cases).

Thanks!
-Suharsh
